Assets/HoloToolkit/Input/Scripts/InputSources/TouchscreenInputSource.cs

```diff
@@ -0,0 +1,238 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using UnityEngine;
+using System.Collections.Generic;
+
+namespace HoloToolkit.Unity.InputModule
+{
+    /// <summary>
+    /// Input source supporting basic touchscreen input:
+    /// * taps
+    /// * holds
+    /// Note that a hold-started is raised as soon as a contact exceeds the Epsilon value;
+    /// if the contact subsequently qualifies as a tap then a hold-cancelled is also raised.
+    /// </summary>
+    public class TouchscreenInputSource : BaseInputSource
+    {
+        const float kContactEpsilon = 2.0f/60.0f;
+
+        [SerializeField]
+        [Tooltip("Time in seconds to determine if the contact registers as a tap or a hold")]
+        protected float MaxTapContactTime = 0.5f;
+
+        private List<PersistentTouch> ActiveTouches;
+
+        private class PersistentTouch
+        {
+            public Touch touchData;
+            public Ray ray;
+            public float lifetime;
+            public PersistentTouch(Touch t, Ray r)
+            {
+                touchData = t;
+                ray = r;
+                lifetime = 0.0f;
+            }
+        }
+
+        #region Unity methods
+
+        protected virtual void Start()
+        {
+            ActiveTouches = new List<PersistentTouch>();
+        }
+
+        protected virtual void Update()
+        {
+            foreach (Touch touch in Input.touches)
+            {
+                // Construct a ray from the current touch coordinates
+                Ray ray = Camera.main.ScreenPointToRay(touch.position);
+
+                switch (touch.phase)
+                {
+                    case TouchPhase.Began:
+                    case TouchPhase.Moved:
+                    case TouchPhase.Stationary:
+                        UpdateTouch(touch, ray);
+                        break;
+
+                    case TouchPhase.Ended:
+                    case TouchPhase.Canceled:
+                        RemoveTouch(touch);
+                        break;
+                }
+            }
+        }
+
+        #endregion // Unity methods
+
+        #region Event generation logic
+
+        public bool UpdateTouch(Touch t, Ray r)
+        {
+            PersistentTouch knownTouch = (ActiveTouches.Find(item => item.touchData.fingerId == t.fingerId));
```

Comment:
Lets optimize this by removing the delegate allocation and just iterating through a for loop